internal/proxy/middleware/obeservability/observability.go

```diff
@@ -0,0 +1,75 @@
+package obeservability
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/odpf/shield/internal/proxy/middleware"
+
+	"github.com/odpf/salt/log"
+	"github.com/rs/xid"
+	"go.uber.org/zap"
+)
+
+const (
+	headerRequestID     = "X-Request-Id"
+	requestIDContextKey = "requestIDContextKey"
+)
+
+type Ware struct {
+	log  *log.Zap
+	next http.Handler
+}
+
+func New(log *log.Zap, next http.Handler) *Ware {
+	return &Ware{
+		log:  log,
+		next: next,
+	}
+}
+
+func (m Ware) Info() *middleware.MiddlewareInfo {
+	return &middleware.MiddlewareInfo{
+		Name:        "_observability",
+		Description: "to handle observability",
+	}
+}
+
+func (m *Ware) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	reqID := GetIncomingRequestID(req)
+	ctx := SetRequestID(req.Context(), reqID)
+	ctx = m.log.NewContext(req.Context())
+	ctx = log.ZapContextWithFields(ctx,
+		zap.String("host", req.Host),
+		zap.String("path", req.URL.String()),
+		zap.String("method", req.Method),
+		zap.String("remote_address", req.RemoteAddr),
+		zap.String("scheme", req.Proto),
+		zap.String("request_id", reqID),
+	)
+	req = req.WithContext(ctx)
+	m.next.ServeHTTP(rw, req)
+}
+
+func GetIncomingRequestID(req *http.Request) string {
```

Comment:
Do these functions need to be public?

Reply:
this can be used from other middleware as well

Reply:
then let's move `GetIncomingRequestID` under `pkg` as a utility function?

Reply:
no, the question is, is this function required outside this package? if not, make it private. otherwise, this is the right place for it. (You can also rename it to `RequestID(_)` by the way. the `GetIncoming` doesn't seem to provide any useful information.

(It's better to avoid `utilities`, `utils`, `common`, etc. as much as possible because they become dumping ground for random things).